de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/console/InviteCommand.java

```diff
@@ -0,0 +1,55 @@
+package de.fu_berlin.inf.dpp.server.console;
+
+import de.fu_berlin.inf.dpp.net.util.XMPPUtils;
+import de.fu_berlin.inf.dpp.net.xmpp.JID;
+import de.fu_berlin.inf.dpp.session.ISarosSessionManager;
+import java.io.PrintStream;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.log4j.Logger;
+
+public class InviteCommand extends ConsoleCommand {
+  private static final Logger log = Logger.getLogger(InviteCommand.class);
+  private final ISarosSessionManager sessionManager;
+
+  public InviteCommand(ISarosSessionManager sessionManager) {
+    this.sessionManager = sessionManager;
+  }
+
+  @Override
+  public String identifier() {
+    return "invite";
+  }
+
+  @Override
+  public String help() {
+    return "invite <JID>... - Invite users to session";
+  }
+
+  @Override
+  public void execute(String command, PrintStream out) {
+    try {
+      List<JID> jids =
+          Arrays.asList(command.split(" "))
+              .stream()
+              .skip(1)
+              .map(x -> new JID(x))
+              .filter(
+                  x -> {
+                    if (!XMPPUtils.validateJID(x)) {
+                      log.warn("Invalid JID skipped: " + x);
```

Comment:
I am about to start sounding pedantic.
Please do not mix concepts. A filter method should only do that: filter.
So, either filter and print afterward or use a forEach method. Do not introduce the concept of side effects to lambdas.
My personal preference: filter and then print
`filter(x -> !XMPPUtils.validateJID(x)).forEach(x -> log.warn("Invalid JID skipped: " + x));`

Reply:
Addition: Yes, that will result in two lists. No, this is not a problem, because we do not have more than 10 JIDs anywhere so there is no performance problem to be seen anywhere.

Reply:
It is not pendantic. The signature is

Stream<T> filter(Predicate<? super T> predicate)

so this implies that there should be no side effects.

Reply:
I don't like the proposed solution as it would mean that we would have to iterate over the JIDs twice.

In general, I wouldn't really think of logging as a side effect. Nonetheless, if this were to be changed, I would definitely prefer moving the check and the logging into a forEach call replacing `filter(...).collect(...);` to avoid additional iterations. This would also mean that we could no longer use `Stream.collect(...)` but would rather have to create an empty collection and add all elements separately.

Reply:
> we would have to iterate over the JIDs twice.
See my comment above. This is not as bad as introducing side effects.

> In general, I wouldn't really think of logging as a side effect.
Oh, but it is.

Reply:
> I am about to start sounding pedantic.
All good, you are right - in principle at least - and I am willing to fix this, no need to excuse.

> I don't like the proposed solution as it would mean that we would have to iterate over the JIDs twice.
But I feel exactly the same as @tobous about this point.

Sure, it would be a micro-optimization, nothing really necessary, but iterating twice is something that bothered me enough to seriously consider switch back to for-loops instead.

I get why side-effects should be avoided and that I am mixing imperative and functional programming styles, but this whole stream is small and very easy to understand.
So yes, I feel this is a little pedantic.

I will check the stream api for a way to partition this jids into two lists (sure there is something), which would at least eliminate the need to call `validateJID` twice for every element.

But this feels a lot like bike-shedding over around 10 lines adding a little debug command.

Reply:
> But this feels a lot like bike-shedding over around 10 lines adding a little debug command.

It might feel that way. But to be honest, I would rather not have the logging of not invited JIDs, than this code. It violates a core programming principle and should be avoided.
There are a lot of discussions about using lambdas and I had enough of these discussions in the last 2 years with people used to using Iterators. If you need performance lambdas are the wrong tool. Heck... Java is the wrong tool!
What you gain from this "double loop": Well speaking code. In this project, the turnover rate is exceptionally high. Except for @srossbach everybody just works a short time on the code. So please be kind to the people who follow and write readable, explicit code with as few side effects as possible.

Reply:
I would also suggest to not optimize code that is mainly just run once. If the application behaves badly performance wise you can just attach a profiler (this is really easy with the JVM). From my experience hotspots normally happen where you do not expect them and the result is normally not from such missing "optimizations" as mentioned here, e.g do not iterate twice but most of the times it is a poor design decision or something else.

I personally seldom use lambas / stream API as they are not helpful and it seems they were tailored to solve enterprise map-reduce algorithms. What is even worse if you meet some guys who are a big fan of functional programming and so ****** up existing easy to read and understandable code by "lambdarizing" everything.

TLDR: If you have the feeling that a lambda expression does solve your problem very cumbersome then do not use it.